pandas/core/dtypes/cast.py

```diff
@@ -331,74 +351,425 @@ def changeit():
+    # * extension arrays
+    # * int vs int
+    # * int vs float / complex
+    # * float vs float
+    # * float vs complex (and vice versa)
+    # * bool
+    # * bytes
+    # * datetimetz
+    # * datetime
+    # * timedelta
+    # * string/object
+
+    # if (is_extension_array_dtype(dtype)
+    #         or is_extension_array_dtype(fill_dtype)):
+    #     # TODO: dispatch to ExtensionDType.maybe_promote? GH 24246
+    if is_integer_dtype(dtype) and is_integer_dtype(fill_dtype):
+        if is_unsigned_integer_dtype(dtype) and all(fill_value >= 0):
+            # can stay unsigned
+            fill_max = fill_value.max()
+            if fill_max > _uint64_max:
+                return np.dtype(object), np.nan
+
+            while fill_max > np.iinfo(dtype).max:
+                # itemsize is the number of bytes; times eight is number of
+                # bits, which is used in the string identifier of the dtype;
+                # if fill_max is above the max for that dtype,
+                # we double the number of bytes/bits.
+                dtype = np.dtype("uint{}".format(dtype.itemsize * 8 * 2))
+            return dtype, None
+        else:
+            # cannot stay unsigned
+            if dtype == "uint64":
+                # need to hold negative values, but int64 cannot hold
+                # maximum of uint64 -> needs object
+                return np.dtype(object), np.nan
+            elif is_unsigned_integer_dtype(dtype):
+                # need to turn into signed integers to hold negative values
+                # int8 cannot hold maximum of uint8; similar for 16/32
+                # therefore, upcast at least to next higher int-type
+                dtype = np.dtype("int{}".format(dtype.itemsize * 8 * 2))
+
+            fill_max = fill_value.max()
+            fill_min = fill_value.min()
+            if isinstance(fill_max, np.uint64):
+                # numpy comparator is broken for uint64;
+                # see https://github.com/numpy/numpy/issues/12525
+                # use .item to get int object
+                fill_max = fill_max.item()
+
+            # comparison mechanics are broken above _int64_max;
+            # use greater equal instead of equal
+            if fill_max >= _int64_max + 1 or fill_min <= _int64_min - 1:
+                return np.dtype(object), np.nan
+
+            while fill_max > np.iinfo(dtype).max or fill_min < np.iinfo(dtype).min:
+                # same mechanism as above, but for int instead of uint
+                dtype = np.dtype("int{}".format(dtype.itemsize * 8 * 2))
+            return dtype, None
+    elif is_integer_dtype(dtype) and is_float_dtype(fill_dtype):
+        # int with float: always upcasts to float64
+        return np.dtype("float64"), np.nan
+    elif is_integer_dtype(dtype) and is_complex_dtype(fill_dtype):
+        # int with complex: always upcasts to complex128
+        return np.dtype("complex128"), np.nan
```

Comment:
should we be distinguishing between np.nan (a float) and e.g. `np.complex128(np.nan)` (a complex128)?

Reply:
I haven't worked much with complex numbers (so I didn't know how `np.complex128(np.nan)` would be cast), but the following would be confusing to me, and it's dangerous IMO to give the false impression that `.imag` is 0.
```
>>> np.complex128(np.nan)
(nan+0j)
>>> np.complex128(np.nan).imag
0.0
```